pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,154 @@
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
+		id64, err := osdStatus.OSD.Int64()
+		if err != nil {
+			continue
+		}
+		id := int(id64)
+
+		logger.Debugf("validating status of osd.%d", id)
+
+		status, _, err := osdDump.StatusByID(int64(id))
+		if err != nil {
+			return err
+		}
+
+		const upStatus int64 = 1
+		var osdDown bool
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.", id)
+			osdDown = false
+			continue
+
+		}
+
+		logger.Debugf("osd.%d is marked 'DOWN'", id)
+
+		// Get the host where the OSD is found
+		hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+		if !osdDown {
+			return fmt.Errorf("OSD %d cannot be removed unless it is 'down'", osdID)
+		}
+
+		// Mark the OSD as out. If it is not down, it is expected this command will fail, in which case
+		// we expect to abort purging the OSD
+		output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
+		if err != nil {
+			return errors.Wrapf(err, fmt.Sprintf("cannot remove the osd. %s", output))
+		}
+
+		// Remove the OSD deployment
+		deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+		logger.Infof("removing the OSD deployment %q", deploymentName)
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if !kerrors.IsNotFound(err) {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+
+		// If running on a PVC, delete the OSD prepare job
+		prepareJobName := fmt.Sprintf("rook-ceph-osd-prepare-example-deviceset-0-#{osdID}")
+		prepareJob, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).Get(prepareJobName, metav1.GetOptions{})
+		logger.Infof("removing the OSD prepare job %q", prepareJobName)
+		if err := k8sutil.DeleteBatchJob(context.Clientset, clusterInfo.Namespace, prepareJobName, false); err != nil {
+			if !kerrors.IsNotFound(err) {
+				// Continue deleting the OSD prepare job even if the deployment fails to be deleted
+				logger.Errorf("failed to delete prepare job for OSD %s. %v", prepareJobName, err)
+			}
+		} else {
+			// If running on a PVC, delete the OSD PVC
+			if pvcName, ok := prepareJob.GetLabels()["ceph.rook.io/pvc"]; ok {
+				// If running on a PVC, delete the OSD PVC
+				logger.Infof("removing the OSD PVC %q", pvcName)
+
+				if err := context.Clientset.CoreV1().PersistentVolumeClaims(clusterInfo.Namespace).Delete(pvcName, &metav1.DeleteOptions{}); err != nil {
+					if !kerrors.IsNotFound(err) {
+						// Continue deleting the OSD PVC even if PVC deletion fails
+						logger.Errorf("failed to delete pvc for OSD %s. %v", pvcName, err)
+					}
+				}
+			}
+
+			// purge the osd
+			output, err = context.Executor.ExecuteCommandWithOutput("ceph", "osd", "purge", fmt.Sprintf("osd.%d", osdID), "--force", "--yes-i-really-mean-it")
```

Comment:
From this point on should not be in the *else* statement. We still want to purge the OSD even if the osd prepare job was not found.